### PR TITLE
feat: add fine-tuning LLM to generate Persian product catalogs in JSON format

### PR DESCRIPTION
**Pull request by MrzEsma**
# What does this PR do?

Hello. In this PR, I tried to create a notebook that fine-tune an LLM as "Hello World" in the simplest way. In addition, to explain the concepts used in this path, I would like to refer to the existing valuable blogs that explained it well and simply. Finally, I also brought a simple code example for fast inference.

## Who can review?

@merveenoyan and @stevhliu

### CODE REVIEW

**Comment by stevhliu** [comment]
Hello, thanks for your contribution!

I think this recipe may be a bit too generic for the cookbook because it doesn't have a specific focus on a real-world application (biomedical, legal, finance domains) or showcase a new technique. One way I think we can improve this recipe is if you show how to finetune the model on domain-specific data.

**Comment by MrzEsma** [comment]
Hello. I didn't realize the focus was on specific domains. Interestingly, I have worked on unstructured data of Iranian products in a well-known Iranian marketplace and was able to obtain JSON output with an LLM. Therefore, I have now modified this Jupyter notebook for this task.
@stevhliu
@merveenoyan

**Comment by stevhliu** [comment]
Cool pivot to showcasing a use case for your recipe! 👏  I think what will make it even more impactful is to generate an example so users can see the output.

**Comment by MrzEsma** [comment]
I deleted the importance level and also wrote an example from the model we had developed on this task to make the output tangible.
@stevhliu
@merveenoyan

File: notebooks/en/fine_tuning_llm_for_generate_persian_product_catalogs_in_json_format.ipynb
Comment (stevhliu) [inline notebooks/en/fine_tuning_llm_for_generate_persian_product_catalogs_in_json_format.ipynb:1]: <p>This should just be regular text instead of a code cell.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/huggingface/cookbook/pull/86/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='3'/>

**Comment by stevhliu** [comment]
One more little thing to fix, otherwise looks good to me!

**Comment by MrzEsma** [comment]
I've moved the example output to a markdown cell as suggested.
@stevhliu

**Review by stevhliu — APPROVED** [review]
LGTM!

File: notebooks/en/fine_tuning_llm_for_generate_persian_product_catalogs_in_json_format.ipynb
Comment (merveenoyan) [inline notebooks/en/fine_tuning_llm_for_generate_persian_product_catalogs_in_json_format.ipynb:1]: <p>Suggestions: "an attempt has been made to fine-tune a Large Language Model (LLM) in the simplest manner possible, without adding unnecessary complexity. " this is more readable: "In this notebook we have attempted to fine-tune a large language model with no added complexity".</p><ul><li>maybe you can give link to VLLM</li><li>No need to explain people, we encourage straightforwardness: I tried to mention all the important concepts and techniques in the description section. However, since others have written valuable explanations and blogs, I simply referred to these sources.</li></ul>
<br/>

 _Reply via <a href='https://app.reviewnb.com/huggingface/cookbook/pull/86/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

Comment (merveenoyan) [inline notebooks/en/fine_tuning_llm_for_generate_persian_product_catalogs_in_json_format.ipynb:1]: <p>so many mix of active and passive voice, avoiding passive voice makes it so readable, maybe you could rewrite for clarity? </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/huggingface/cookbook/pull/86/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

Comment (merveenoyan) [inline notebooks/en/fine_tuning_llm_for_generate_persian_product_catalogs_in_json_format.ipynb:1]: <p>These a re hyperparameters, maybe you can remove comments and explain what each block is for in a markdown cell here and add comments for granular ones, i.e. walk reader through what you're doing and how you're doing it? </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/huggingface/cookbook/pull/86/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='5'/>

Comment (merveenoyan) [inline notebooks/en/fine_tuning_llm_for_generate_persian_product_catalogs_in_json_format.ipynb:1]: <p>maybe move this up ⬆️ where you're defining LoRA hyperparameters. </p><p>also for LoRA I think it's worth mentioning you are storing lower rank matrices that store changes in weights since it's the main point.</p><p>If you want you can always provide further reading at the bottom of the recipe not to divert the attention.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/huggingface/cookbook/pull/86/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='9'/>

Comment (merveenoyan) [inline notebooks/en/fine_tuning_llm_for_generate_persian_product_catalogs_in_json_format.ipynb:1]: <p>you can put this in the bottom</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/huggingface/cookbook/pull/86/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='13'/>

Comment (merveenoyan) [inline notebooks/en/fine_tuning_llm_for_generate_persian_product_catalogs_in_json_format.ipynb:1]: <p>you can move this up ⬆️ above the chat template, also we are using I and we interchangeably, best to stick to we.</p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/huggingface/cookbook/pull/86/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='17'/>

Comment (merveenoyan) [inline notebooks/en/fine_tuning_llm_for_generate_persian_product_catalogs_in_json_format.ipynb:1]: <p>Maybe explain SFTTrainer just in one sentence, people might not know and some params like packing or formatting function that we normally don't have in Trainer</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/huggingface/cookbook/pull/86/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='19'/>

Comment (merveenoyan) [inline notebooks/en/fine_tuning_llm_for_generate_persian_product_catalogs_in_json_format.ipynb:1]: <p>you can merge this to the cell above</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/huggingface/cookbook/pull/86/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='23'/>

Comment (merveenoyan) [inline notebooks/en/fine_tuning_llm_for_generate_persian_product_catalogs_in_json_format.ipynb:1]: <p>newlines in functions help with readability</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/huggingface/cookbook/pull/86/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='24'/>

Comment (merveenoyan) [inline notebooks/en/fine_tuning_llm_for_generate_persian_product_catalogs_in_json_format.ipynb:1]: <p>You could explain that adapter models need to be merged to base model but here you're pushing the model too, this needs mentioning</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/huggingface/cookbook/pull/86/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='29'/>

**Comment by MrzEsma** [comment]
I tried to follow all the points. If there is anything else, tell me.
@merveenoyan

**Comment by MrzEsma** [comment]
Is it ok?

**Comment by MrzEsma** [comment]
Is it ok?
@merveenoyan

**Comment by MrzEsma** [comment]
Hello. I am still waiting for your approval and merge, or any feedback you may have. If there are any issues, please let me know so I can fix them.
@merveenoyan

**Comment by MrzEsma** [comment]
Hello. Does this PR still have a point to merge? If you tell me, I will fix it. Because no matter what I asked Mrs. merveenoyan, she did not answer.
@stevhliu

File: notebooks/en/fine_tuning_llm_for_generate_persian_product_catalogs_in_json_format.ipynb
Comment (stevhliu) [reply notebooks/en/fine_tuning_llm_for_generate_persian_product_catalogs_in_json_format.ipynb:1]: <p>You could actually remove this entire paragraph since this notebook doesn't show how to do alignment.</p>

Comment (stevhliu) [reply notebooks/en/fine_tuning_llm_for_generate_persian_product_catalogs_in_json_format.ipynb:1]: <p>I don't think this is resolved yet. You can break up this large cell block into smaller blocks and give a brief description of what's happening inside each block.</p>

Comment (stevhliu) [reply notebooks/en/fine_tuning_llm_for_generate_persian_product_catalogs_in_json_format.ipynb:1]: <p>Not resolved yet. I think Merve meant putting it at the bottom of the notebook so readers don't get distracted.</p>

**Review by stevhliu — COMMENTED** [review]
Sorry for the delay!

I left a few more comments about some unresolved issues.

**Comment by MrzEsma** [comment]
I've tried to incorporate your feedback and address the issues. If there are still any problems, please let me know.

@stevhliu

File: notebooks/en/fine_tuning_llm_to_generate_persian_product_catalogs_in_json_format.ipynb
Comment (merveenoyan) [inline notebooks/en/fine_tuning_llm_to_generate_persian_product_catalogs_in_json_format.ipynb:1]: <p>Suggestion: *Model Training</p><p> </p><p>Also would be nice to walk people through what you're doing when you load bnb config, lora config and more (we try to assume the reader isn't very knowledgeable)</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/huggingface/cookbook/pull/86/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='11'/>

**Comment by merveenoyan** [comment]
@MrzEsma Hello, I just left a very general comment, we can merge once addressed :)

**Comment by MrzEsma** [comment]
hello. I tried to review the notebook again and explain any concept that is specific to fine-tuning language models. Once again, if you think there's something I haven't explained, please mention it. Thank you.
@merveenoyan
@stevhliu

**Pull request merged**
